coil-base/src/main/java/coil/decode/BitmapFactoryDecoder.kt

```diff
@@ -78,7 +79,15 @@ class BitmapFactoryDecoder @JvmOverloads constructor(
         // Always create immutable bitmaps as they have performance benefits.
         inMutable = false
 
-        if (outWidth > 0 && outHeight > 0) {
+        if (options.size.isOriginal && source.metadata is ResourceMetadata) {
```

Comment:
This won't account for the case when only one dimension is `Dimension.Original`; also I'm not keen on adding a special case in `BitmapFactoryDecoder` only for resources. Maybe we only need to set the density after the bitmap is decoded and we can avoid creating a special case in `BitmapFactoryDecoder`?

Reply:
_This won't account for the case when only one dimension is Dimension.Original_
Good callout when one dimension is `Dimension.Original`. Do you think we need to support mixed `Dimension.Original` with `Dimension.Pixel`? Is this something we can regard as invalid?

_Also do we need to make similar changes in the other decoders_
I would need to check the behavior of `ImageDecoderDecoder`. Other decoders are probably not density-aware.

_Is it possible to avoid adding special cases to each decoder? Maybe we can set the density after the bitmap is decoded and avoid creating a special case in BitmapFactoryDecoder?_
1. The problem with resetting the density after `BitmapFactory.decode()` is that, we probably need another bitmap transformation to scale the previous image to meet the dimension need.
2. For additional special case, it might be necessary because Android `BitmapFactory` and `ImageDecoder` both expose specific API to decode native `Resource`. `Resource` is unique from other sources in that it is density-aware.

Reply:
I played around with `ImageDecoderDecoder` and `GifDecoder`. They seem to be very different from the default `BitmapFactory` decoder - Therefore this PR might be better to scoped to only handle `Bitmap`. I have updated the PR title accordingly.

Reply:
>Do you think we need to support mixed Dimension.Original with Dimension.Pixel?

My colleague @CordulaMW created the issue #1053 for this PR, so I can comment on at least what we need: We don't need mixed Dimension values. We would either use `Dimension.Original` or `Dimension.Pixel`, never both at the same time.  

I generally don't think it would make sense to support this, as it's afaik only for scaling. Why would someone want the height dpi-scaled but the width not?   
As a counter argument, the android layout system does support it. Not sure.